test/src/test/java/org/corfudb/integration/StreamingIT.java

```diff
@@ -441,6 +443,130 @@ public void testStreamingMultiTableSameListener() throws Exception {
         assertThat(shutdownCorfuServer(corfuServer)).isTrue();
     }
 
+    /**
+     * Example to show how to retrieve the previous value from a streamed entry
+     * @param <K>
+     * @param <V>
+     * @param <M>
+     */
+    class PrevValueStreamer<K extends Message, V extends Message, M extends Message> implements StreamListener {
+        @Getter
+        private int recordCount;
+        private CorfuStore corfuStore;
+        private final String namespace;
+        private final String tableName;
+        public PrevValueStreamer(CorfuStore corfuStore, String namespace, String tableName) {
+            recordCount = 0;
+            this.corfuStore = corfuStore;
+            this.namespace = namespace;
+            this.tableName = tableName;
+        }
+
+        public String toString() {
+            return namespace+tableName;
+        }
+        /**
+         * A corfu update can/may have multiple updates belonging to different streams.
+         * This callback will return those updates as a list grouped by their Stream UUIDs.
+         *
+         * @param results is a map of stream UUID -> list of entries of this stream.
+         */
+        @Override
+        public void onNext(CorfuStreamEntries results) {
+            results.getEntries().forEach((k, v) -> {
+                v.forEach(entry -> {
+                    final CorfuStoreEntry<K, V, M> previousRecord =
+                            getPreviousRecord(namespace, tableName, entry, entry.getEpoch(), entry.getAddress());
+                    SampleTableAMsg prevMsg = (SampleTableAMsg) previousRecord.getPayload();
+                    if (recordCount > 0) {
+                        assertThat(prevMsg.getPayload()).isEqualTo("val"+(recordCount - 1));
+                    } else {
+                        assertThat(previousRecord.getPayload()).isNull();
+                    }
+                });
+            });
+            recordCount++;
+        }
+
+        /**
+         *  Fetch the previous value of a record from the table using the sequence - 1 snapshot
+         * @param namespace
+         * @param tableName
+         * @param entry
+         * @param epoch
+         * @param sequence
+         * @return
+         */
+        public CorfuStoreEntry<K, V, M> getPreviousRecord(String namespace, String tableName,
+                                                          CorfuStreamEntry entry,
+                                                          long epoch,
+                                                          long sequence) {
+            Timestamp prevTimestamp = Timestamp.newBuilder()
+                    .setEpoch(epoch)
```

Comment:
For the actual user, is it possible that the previous timestamp has a lower epoch? If so, will it get previous record or a null value?

Reply:
this entry carries the exact epoch and address, so if the previous timestamp epoch is lower, then it should get the previous record. I haven't tested that case though to be sure. thanks